### PR TITLE
Fixes and improvements for the TKDE class

### PR DESCRIPTION
**Pull request by lmoneta**
Apply some bug fixes in the TKDE class.
 - Fix initialization for the automatic binning case
 - Fix setting the bins using the setter functions.
 - Fix mirroring for binned case (fix ROOT-8488)
 - Fix MirrorAsym case 
 - Fix computation of non-adaptive bandwidth for binned case
  
Add also some improvements in the class avoid some unneeded re-initialization and some small speedup in the kernel calculations.
Improve also the documentations

### CODE REVIEW

File: hist/hist/inc/TKDE.h [hunk 1]
```diff
@@ -42,46 +46,63 @@ class TKDE : public TNamed  {
...
       Instantiate( nullptr,  events, data, nullptr, xMin, xMax, option, rho);
    }
 
+   /// Constructor for weighted data
    TKDE(UInt_t events, const Double_t* data, const Double_t* dataWeight, Double_t xMin = 0.0, Double_t xMax = 0.0, const Option_t* option =
         "KernelType:Gaussian;Iteration:Adaptive;Mirror:noMirror;Binning:RelaxedBinning", Double_t rho = 1.0) {
       Instantiate( nullptr,  events, data, dataWeight, xMin, xMax, option, rho);
    }
 
+   /// Constructor for unweweighted data and a user defined kernel function
```
Comment (guitargeek) [inline hist/hist/inc/TKDE.h:100]: There is a typo: `unweweighted`

File: hist/hist/inc/TKDE.h [hunk 2]
```diff
@@ -42,46 +46,63 @@ class TKDE : public TNamed  {
...
    template<class KernelFunction>
    TKDE(const Char_t* /*name*/, const KernelFunction& kernfunc, UInt_t events, const Double_t* data, Double_t xMin = 0.0, Double_t xMax = 0.0, const Option_t* option = "KernelType:UserDefined;Iteration:Adaptive;Mirror:noMirror;Binning:RelaxedBinning", Double_t rho = 1.0)  {
       Instantiate(new ROOT::Math::WrappedFunction<const KernelFunction&>(kernfunc), events, data, nullptr, xMin, xMax, option, rho);
    }
+   /// Constructor for weweighted data and a user defined kernel function
```
Comment (guitargeek) [inline hist/hist/inc/TKDE.h:105]: There is a typo: `weweighted`

File: hist/hist/inc/TKDE.h (hunk 1 above)
Comment (guitargeek) [reply hist/hist/inc/TKDE.h:100]: By the way, is the documentation automatically checked for typos at some point? Just to know if it's actually necessary to sporadically pay attention to spelling in PR reviews.

File: hist/hist/src/TKDE.cxx [hunk 3]
```diff
@@ -563,22 +595,38 @@ void TKDE::SetSigma(Double_t R) {
...
+   // it should not be fData.size() that in binned case is number of bins
+   UInt_t n =  (fUseBins) ? fNBins : fNEvents;
    if (n == 0) return;
    // Optimal bandwidth (Silverman's rule of thumb with assumed Gaussian density)
    Double_t weight = fCanonicalBandwidths[kGaussian] * fSigmaRob * std::pow(3. / (8. * std::sqrt(M_PI)) * n, -0.2);
    weight *= fRho * fCanonicalBandwidths[fKernelType] / fCanonicalBandwidths[kGaussian];
-   if (fKernel) delete fKernel;
-   fKernel = new TKernel(weight, this);
+
+   fKernel.reset( new TKernel(weight, this) );
```
Comment (guitargeek) [inline hist/hist/src/TKDE.cxx:606]: I think in this situation, `fKernel = std::make_unique<TKernel>(weight, this)` should be preferred. At least that's what I read in the core guidelines:

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#r23-use-make_unique-to-make-unique_ptrs

Comment (lmoneta) [reply hist/hist/src/TKDE.cxx:606]: I agree, make_unique is preferable, but I think is from C++-14 and not 11.

File: hist/hist/inc/TKDE.h (hunk 1 above)
Comment (lmoneta) [reply hist/hist/inc/TKDE.h:100]: Thank you for checking the spelling. Unfortunately the documentation does not check for this automatically

**Review by guitargeek — APPROVED** [review]
Looks good to me!

**Pull request merged**